In [2]:
import torch

# 初始化权重参数
w = torch.tensor([1.0], requires_grad=True,dtype=torch.float32)
# 实例化优化器
optimizer = torch.optim.RMSprop([w],lr=0.01,alpha=0.9)
# 迭代的次数
count =3
for i in range(count):
    # 目标函数
    y = (w**2)/2
    # 清空梯度
    optimizer.zero_grad()
    # 计算梯度
    y.backward()
    # 更新参数
    optimizer.step()
    # 打印结果
    print(f'梯度:{w.grad.item():.3f}, 权重:{w.item():.3f}')

梯度:1.000, 权重:0.968
梯度:0.968, 权重:0.946
梯度:0.946, 权重:0.927


### 最终输出

综合每次迭代的计算，输出如下（保留 3 位小数）：

```
梯度:1.000, 权重:0.968
梯度:0.968, 权重:0.946
梯度:0.946, 权重:0.927
```

---

### 代码背景

- 目标函数：$ y = \frac{w^2}{2} $
- 梯度：$ \frac{dy}{dw} = w $
- 优化器：RMSprop，学习率 `lr=0.01`，动量因子 `alpha=0.9`，默认 $\epsilon=1e-8$（PyTorch 默认值）
- 初始权重：$ w = 1.0 $
- 迭代次数：3 次

RMSprop 更新公式：

1. 计算梯度 $ g_t = \frac{dy}{dw} $

2. 更新梯度平方的指数移动平均：
   $$
   v_t = \alpha \cdot v_{t-1} + (1 - \alpha) \cdot g_t^2
   $$

3. 更新参数：
   $$
   w_t = w_{t-1} - \frac{\text{lr}}{\sqrt{v_t + \epsilon}} \cdot g_t
   $$

初始条件：

- $ w_0 = 1.0 $
- $ v_0 = 0 $（PyTorch 初始化时，均方梯度为 0）

---

### 迭代计算过程

#### **第一次迭代**

1. **计算目标函数和梯度**：

   - 权重：$ w = 1.0 $
   - 目标函数：$ y = \frac{w^2}{2} = \frac{1.0^2}{2} = 0.5 $
   - 梯度：$ g_1 = \frac{dy}{dw} = w = 1.0 $

2. **RMSprop 更新**：

   - 计算梯度平方的移动平均：
     $$
     v_1 = \alpha \cdot v_0 + (1 - \alpha) \cdot g_1^2 = 0.9 \cdot 0 + (1 - 0.9) \cdot 1.0^2 = 0.1
     $$

   - 计算更新步长：
     $$
     \text{step} = \frac{\text{lr}}{\sqrt{v_1 + \epsilon}} \cdot g_1 = \frac{0.01}{\sqrt{0.1 + 1e-8}} \cdot 1.0
     $$

     $$
     \sqrt{0.1} \approx 0.316227766, \quad \sqrt{0.1 + 1e-8} \approx 0.316227766
     $$

     $$
     \text{step} \approx \frac{0.01}{0.316227766} \cdot 1.0 \approx 0.0316227766
     $$

   - 更新权重：
     $$
     w_1 = w_0 - \text{step} = 1.0 - 0.0316227766 \approx 0.9683772234
     $$

3. **输出**：

   - 梯度：$ g_1 = 1.0 $
   - 权重：$ w_1 \approx 0.968 $
   - 打印：`梯度:1.000, 权重:0.968`

---

#### **第二次迭代**

1. **计算目标函数和梯度**：

   - 权重：$ w \approx 0.9683772234 $
   - 目标函数：$ y = \frac{w^2}{2} = \frac{0.9683772234^2}{2} \approx 0.468754468 $
   - 梯度：$ g_2 = \frac{dy}{dw} = w \approx 0.9683772234 $

2. **RMSprop 更新**：

   - 计算梯度平方的移动平均：
     $$
     v_2 = \alpha \cdot v_1 + (1 - \alpha) \cdot g_2^2
     $$

     $$
     g_2^2 \approx 0.9683772234^2 \approx 0.937754468
     $$

     $$
     v_2 = 0.9 \cdot 0.1 + (1 - 0.9) \cdot 0.937754468 \approx 0.09 + 0.0937754468 \approx 0.1837754468
     $$

   - 计算更新步长：
     $$
     \text{step} = \frac{\text{lr}}{\sqrt{v_2 + \epsilon}} \cdot g_2 = \frac{0.01}{\sqrt{0.1837754468 + 1e-8}} \cdot 0.9683772234
     $$

     $$
     \sqrt{0.1837754468} \approx 0.428690896, \quad \sqrt{0.1837754468 + 1e-8} \approx 0.428690896
     $$

     $$
     \text{step} \approx \frac{0.01}{0.428690896} \cdot 0.9683772234 \approx 0.0233216708 \cdot 0.9683772234 \approx 0.0225823703
     $$

   - 更新权重：
     $$
     w_2 = w_1 - \text{step} \approx 0.9683772234 - 0.0225823703 \approx 0.9457948531
     $$

3. **输出**：

   - 梯度：$ g_2 \approx 0.968 $
   - 权重：$ w_2 \approx 0.946 $
   - 打印：`梯度:0.968, 权重:0.946`

---

#### **第三次迭代**

1. **计算目标函数和梯度**：

   - 权重：$ w \approx 0.9457948531 $
   - 目标函数：$ y = \frac{w^2}{2} = \frac{0.9457948531^2}{2} \approx 0.447524566 $
   - 梯度：$ g_3 = \frac{dy}{dw} = w \approx 0.9457948531 $

2. **RMSprop 更新**：

   - 计算梯度平方的移动平均：
     $$
     v_3 = \alpha \cdot v_2 + (1 - \alpha) \cdot g_3^2
     $$

     $$
     g_3^2 \approx 0.9457948531^2 \approx 0.894527132
     $$

     $$
     v_3 = 0.9 \cdot 0.1837754468 + (1 - 0.9) \cdot 0.894527132 \approx 0.1653979021 + 0.0894527132 \approx 0.2548506153
     $$

   - 计算更新步长：
     $$
     \text{step} = \frac{\text{lr}}{\sqrt{v_3 + \epsilon}} \cdot g_3 = \frac{0.01}{\sqrt{0.2548506153 + 1e-8}} \cdot 0.9457948531
     $$

     $$
     \sqrt{0.2548506153} \approx 0.504823262, \quad \sqrt{0.2548506153 + 1e-8} \approx 0.504823262
     $$

     $$
     \text{step} \approx \frac{0.01}{0.504823262} \cdot 0.9457948531 \approx 0.0198071537 \cdot 0.9457948531 \approx 0.0187353456
     $$

   - 更新权重：
     $$
     w_3 = w_2 - \text{step} \approx 0.9457948531 - 0.0187353456 \approx 0.9270595075
     $$

3. **输出**：

   - 梯度：$ g_3 \approx 0.946 $
   - 权重：$ w_3 \approx 0.927 $
   - 打印：`梯度:0.946, 权重:0.927`